### PR TITLE
add an abstract output layer.

### PR DESCRIPTION
**Pull request by shinnya**
To abstract output layer, add `Writer` interface and `StdoutWriter`, which is an implementation class of `Writer`.

### CODE REVIEW

**Comment by shinnya** [comment]
Are these changes like what you imagined?

**Comment by c9s** [comment]
Maybe we just rename `StdoutWriter` to StdWriter` so we can use stdout and stderr together.

**Comment by c9s** [comment]
ah, I think it makes sense. but can we also provide StdErrWriter ?

**Comment by c9s** [comment]
And one more method:  `public function writef($format, $args...)`

**Comment by shinnya** [comment]
Ok, I got it.

**Comment by shinnya** [comment]
We can output to stderr by `fwrite(STDERR, "stderr\n");`

**Comment by shinnya** [comment]
Now we can use `EchoWriter` for unit tests and `StreamWriter` for outputing to `STDOUT` or `STDERR`. `ServiceContainer` uses `StreamWriter(STDOUT)`.

**Comment by c9s** [comment]
Looks great, let's see why travis-ci is failing?

**Comment by c9s** [comment]
Looks like the output is not caught for testing? we have to make sure that both bash/zsh completion output the strings in stdout. :)

**Comment by shinnya** [comment]
> Looks great, let's see why travis-ci is failing?
> Looks like the output is not caught for testing?

Thank you :smile: The reason why some tests were failing was that I didn't modify tests for DemoApp, which wrote output to console.

File: src/CLIFramework/ServiceContainer.php [hunk 1]
```diff
@@ -39,7 +44,11 @@ public function __construct()
     static public function getInstance()
     {
         static $instance;
-        $instance = new self;
```
Comment (shinnya) [inline src/CLIFramework/ServiceContainer.php]: This is a bug that calling getInstancealways creates a new instance.

**Comment by c9s** [comment]
great! thanks!

**Pull request merged**